### ISSUE TITLE
Vaultparameters hashtable should be case insensitive.

### ISSUE THREAD

**Issue by itfranck** [PowerShell/SecretManagement#69]
Most of Powershell is case insensitive. 
VaultParameters is taking in a Powershell hashtable but it is not case insensitive.

This will succed: 
```
Register-SecretVault -ModuleName Az.Keyvault -VaultParameters @{ AZKVaultName= 'devops-cicd'; SubscriptionId = '<ID>' }
```

This will fail: 
```
Register-SecretVault -ModuleName Az.Keyvault -VaultParameters @{ AZKVaultName= 'devops-cicd'; subscriptionId = '<ID>' }
```
will fail. 

(If you're looking for the issue, "Subscriptionid" in the failing registration have its first 'S' in lower case. 

```

# .Net C# code for SecretManagement
$SensitiveHT = [hashtable]::new()
$SensitiveHT.HELLO = 'WORLD'

#Return Null  :( 
$SensitiveHT.hello 

$PowershellHT = @{'HELLO'='WORLD'}
# Do not care about what anyone think (Insensitive)
# Return WORLD
$PowershellHT.hello 

# Expected .Net C# hashtable generation
$InsensitiveHT = [hashtable]::new([System.StringComparer]::CurrentCultureIgnoreCase)
$InsensitiveHT.HELLO = 'WORLD'

#Return WORLD
$InsensitiveHT.hello
```

**Comment by PaulHigin** [comment]
I think it makes sense to have additional parameters be case sensitive in this situation.  This is by design.

**Comment by itfranck** [comment]
The value is case-sensitive or case-insensitive depending on the implementation and I agree it can make sense since the system behind it might expect a certain case.

The name though ?
The case name of the parameter should be unimportant and received by the implementing module regardless of the case. 

I cannot imagine a scenario where you would purposefully want a key name case-sensitive and deal with 
`@{SUBSCRIPTIONID = 'Value' }` and `@{subsciprionID = 'Value' }` in a distinct way. 
Usually Powershell is very loose on that aspect (and PS hashtables are all insensitive) 

I took some time to debug an issue on an implementation not catching on that the hashtable key name case was relevant in the context of SecretManagement (as when dealing with pure powersehll, it is irrelevant) 
Anyhow, this was just a simple suggestion based on usage. 
It is definitely not deal breaking.

**Comment by PaulHigin** [comment]
SecretManagement just passes these name/value pairs to the extension vault implementation.  It is up to the individual implementation to enforce case sensitivity.

**Comment by itfranck** [comment]
> SecretManagement just passes these name/value pairs to the extension vault implementation. It is up to the individual implementation to enforce case sensitivity.

Exact.
I do not debate that.
What I was debating was more the communication of these values in-between the user script and the implemented module.
The implemented module is not able to retrieve the key if the case do not match, which can be understood when you know there's another stack (C#) in the middle doing a bridge between the user script and the implemented module but from a pure Powershell perspective, the hashtable key would normally be retrievable in a case-insensitive way rather than being case dependant. 

This wouldn't affect how the implementing module deal with the hashtable afterward and whether or not it implement case sensitive restriction on its end but simply allow getting the value of the vault parameters in the module even if the case of the hashtable key was not respected. 

Here's an example that demonstrate both case (Powershell vs simulation of C# module in the middle using case-sensitive hashtable) 

```

# Pure powershell hashtable
Function Register-SecretVault2($VaultParameters) {
    ImplementedModulePSM1  -AdditionalParameters $VaultParameters

}

# SecretManagement in the middle, loose the original case insensitive hashtable properties
Function Register-SecretVault3($VaultParameters) {
    #C# hashtable
    $hash = [system.collections.hashtable]::new()
    $VaultParameters.keys | % {$hash.Add($_,$VaultParameters[$_])}
    ImplementedModulePSM1  -AdditionalParameters $hash

}


#Simulate the implemented module and how I can check if user provided a specific parameter key
Function ImplementedModulePSM1($AdditionalParameters) {
    return $AdditionalParameters.subscriptionid
}

# Case for the key do not matter, unless implemented module choose to do something with key /value pair 
Write-Host "The Powershell case insensitive way " -ForegroundColor Cyan
$Value = Register-SecretVault2 -VaultParameters @{SUBSCRIPTIONID = '12345' }
Write-Host "Value was null: $($null -eq $Value)" -ForegroundColor Green
Write-Host "Value: $Value" -ForegroundColor Green

Write-Host "The current way, which won't detect a valid parameter key of the wrong case" -ForegroundColor Cyan
$Value = Register-SecretVault3 -VaultParameters @{SUBSCRIPTIONID = '12345' }
Write-Host "Value was null: $($Value -eq $null)" -ForegroundColor Yellow
```

tl;dr 
Powershell deal with hashtable in a case-insensitive way by default.
C# deal with hashtables in a case-sensitive way by default.
Since this module is a bridge between Powershell and Powershell, it should preserve the behaviors of Powershell rather than impose its own case-sensitivity restriction into the mix.

(Again, this is not a big issue, this was just something unusual that caused some issue and had me add forced verbose to determine that case sensitivity was lost in the process, which I was not expecting as I was dealing with a powershell script and a powershell implemented module. I just wanted to make sure it was understood the exact reason why I opened this ticket in the first place and this is not too much of a worry, I'll just remember this specific consideration when implementing new module / using those.)
